src/libraries/System.ServiceProcess.ServiceController/tests/System.ServiceProcess.ServiceController.TestService/TestServiceInstaller.cs

```diff
@@ -174,17 +180,24 @@ private void DeleteService()
             using (var serviceManagerHandle = new SafeServiceHandle(Interop.Advapi32.OpenSCManager(null, null, Interop.Advapi32.ServiceControllerOptions.SC_MANAGER_ALL)))
             {
                 if (serviceManagerHandle.IsInvalid)
-                    throw new Win32Exception("Could not open SCM");
+                {
+                    string errorMessage = new Win32Exception().Message;
+                    throw new Win32Exception($"Could not open SCM. {errorMessage}");
+                }
 
                 using (var serviceHandle = new SafeServiceHandle(Interop.Advapi32.OpenService(serviceManagerHandle, ServiceName, Interop.Advapi32.ServiceOptions.STANDARD_RIGHTS_DELETE)))
                 {
                     if (serviceHandle.IsInvalid)
-                        throw new Win32Exception($"Could not find service '{ServiceName}'");
+                    {
+                        string errorMessage = new Win32Exception().Message;
+                        throw new Win32Exception($"Could not find service '{ServiceName}'. {errorMessage}");
+                    }
 
                     TestService.DebugTrace("TestServiceInstaller: instructing ServiceController to Delete service " + ServiceName);
                     if (!Interop.Advapi32.DeleteService(serviceHandle))
                     {
-                        throw new Win32Exception($"Could not delete service '{ServiceName}'");
+                        string errorMessage = new Win32Exception().Message;
+                        throw new Win32Exception($"Could not delete service '{ServiceName}'. {errorMessage}");
```

Comment:
This ends up calling https://github.com/dotnet/runtime/blob/441271c95fe6e4d2b4bada7deb037a44d0eb8c74/src/libraries/System.Private.CoreLib/src/System/ComponentModel/Win32Exception.cs#L46 that will read the last PInvoke error that can be corrupted by now.

Reply:
Ugh - Win32Exception is a bit of a trap. So is this the correct pattern to achieve this?
```c#
Win32Exception ex = new Win32Exception();
string errorMessage= ex.Message;
string code = ex.NativeErrorCode;
throw new Win32Exception(code, $"Could not delete service '{ServiceName}'. {errorMessage}");
```

Reply:
Yes, that is correct. I think writing it like this looks better:
```
int errorCode = Marshal.GetLastWin32Error();
string errorMessage = new Win32Exception(errorCode).Message;
throw new Win32Exception(errorCode, $"Could not delete service '{ServiceName}'. {errorMessage}");
```

Reply:
similar
https://github.com/danmoseley/runtime/blob/a431bbc76dd80cb6f960725547ad3ccb5c9cf679/src/libraries/System.Drawing.Common/src/misc/DbgUtil.cs#L57-L59

I wonder what would be an appropriate new API to make this (seemingly reasonable) pattern easier.
```
throw Win32Exception.WithMessagePrefix($"Could not delete service '{ServiceName}.');
```
?

Reply:
Actually, the above suggestion would still run string formatting machinery before reading the last error.

Reply:
I think an API that forces you to call `Marshal.GetLastWin32Error` yourself and gives you the error message for the OS error code would be more flexible.

I think the current behavior of Win32Exceptions that calls `Marshal.GetLastWin32Error` for you is very error prone.

Reply:
Hmm, so eg this?  (assuming I want the code in the exception object too)
```c#
int errorCode = Marshal.GetLastPinvokeError();
int errorMessage = Marshal.GetMessageForPinvokeError(errorCode);
throw new Win32Exception(errorCode, $"Could not delete service '{ServiceName}'. {errorMessage}");
```

Reply:
created #67872